### PR TITLE
add HELMET_COMPAT flag, allow one headgear at once

### PR DESCRIPTION
**Pull request by ymber**
Changes the conditions for allowing headgear to be worn so that one main headgear can be worn with layers underneath it and other compatible headgear. This system is clearer about what headgear can be worn at the same time than the current encumbrance check and makes it easier to check whether a helmet is currently worn.

Balaclava(SKINTIGHT) + chainmail(HELMET_COMPAT) + helmet should be acceptable. Balaclava(SKINTIGHT) + helmet + beret should not. Under this implementation the second is disallowed automatically and the first is allowed with a HELMET_COMPAT flagged chainmail coif.

### CODE REVIEW

**Comment by smolbird** [comment]
No JSON changes to test that the flag works?

**Comment by ymber** [comment]
I tested locally and it worked as intended. I just made a test item for that and I want to do all the tiems that need HELMET_COMPAT in one go.

File: src/player.cpp [hunk 1]
```diff
@@ -10551,6 +10550,21 @@ bool player::is_wearing_shoes(std::string side) const
     return (left && right);
 }
 
+bool player::is_wearing_helmet() const
+{
+    bool ret = false;
+    for (auto &i : worn) {
```
Comment (illi-kun) [inline src/player.cpp]: This should be properly formatted per current astyle rules used on the project (even if the file you working with is not yet properly formatted).

**Comment by illi-kun** [comment]
@ymber
At least one item should be updated for two reasons:
- for testing purpose
- for demonstration of way of use a new feature.

**Comment by ymber** [comment]
Added HELMET_COMPAT to some appropriate items. Going through the list of head covering stuff in armor.json makes me think OVERSIZE items should be exempt from helmet exclusivity as well.

**Comment by kevingranade** [comment]
New status is that you can wear ANY amount of HELMET_COMPAT items at once, I think we need to reign that in some.
Simplest thing seems to be to bring back a encumbrance total, but perhaps make it a larger number.
i.e. "you can wear up to 20 enc of SKINTIGHT + HELMET_COMPAT under* your helmet, and any amount of OVERSIZE items over* your helmet".

*Not checking if it's over/under, just assuming that's where they go.

File: src/player.cpp [hunk 2]
```diff
@@ -7676,6 +7676,12 @@ ret_val<bool> player::can_wear( const item& it  ) const
         return ret_val<bool>::make_failure( wearing_something_on( bp_head ), _( "You can't wear that with other headgear!" ) );
     }
 
+    if( it.covers( bp_head ) &&
+      ( it.has_flag( "SKINTIGHT" ) || it.has_flag( "HELMET_COMPAT" ) ) &&
+      ( head_cloth_encumbrance() + it.get_encumber() >= 20 ) ) {
+        return ret_val<bool>::make_failure( _( "You cant wear that much on your head!" ) );
```
Comment (Night-Pryanik) [inline src/player.cpp]: "cant" -> "can't"

File: doc/JSON_FLAGS.md [hunk 3]
```diff
@@ -709,6 +709,7 @@ Some armor flags, such as `WATCH` and `ALARMCLOCK` are compatible with other ite
 - ```RAINPROOF``` Prevents the covered body-part(s) from getting wet in the rain.
 - ```RESTRICT_HANDS``` Prevents the player from wielding a weapon two-handed, forcing one-handed use if the weapon permits it.
 - ```SKINTIGHT``` Undergarment layer.
+- ```HELMET_COMPAT``` Items that are not SKINTIGHT or OVERSIZE but can be worn with a helmet.
```
Comment (Night-Pryanik) [inline doc/JSON_FLAGS.md]: Please place your tag in alphabetical order.

**Comment by matskuman5** [comment]
Shouldn't it be HELMET_COMPACT?

**Comment by Night-Pryanik** [comment]
>Shouldn't it be HELMET_COMPACT?

I think COMPAT is the abbreviation of COMPATIBLE or COMPATIBILITY.

**Comment by ymber** [comment]
Yeah, compatible.

File: src/player.cpp [hunk 4]
```diff
@@ -7672,9 +7672,14 @@ ret_val<bool> player::can_wear( const item& it  ) const
         return ret_val<bool>::make_failure( _( "You're already wearing footwear!" ) );
     }
 
-    if( it.covers( bp_head ) && ( encumb( bp_head ) > 10) && ( !( it.get_encumber() < 9) ) ) {
-        return ret_val<bool>::make_failure( wearing_something_on( bp_head ) ?
-                            _( "You can't wear another helmet!" ) : _( "You can't wear a helmet!" ) );
+    if( it.covers( bp_head ) && !it.has_flag( "HELMET_COMPAT" ) && is_wearing_helmet() ) {
```
Comment (kevingranade) [inline src/player.cpp]: This first check will bail out on SKINTIGHT items when it shouldn't.

File: src/player.cpp [hunk 5]
```diff
@@ -10551,6 +10556,34 @@ bool player::is_wearing_shoes(std::string side) const
     return (left && right);
 }
 
+bool player::is_wearing_helmet() const
+{
+    bool ret = false;
+    for( auto &i : worn ) {
+        const item *worn_item = &i;
```
Comment (Coolthulhu) [inline src/player.cpp]: You don't need this pointer at all, just use `i`.

File: src/player.cpp [hunk 6]
```diff
@@ -10551,6 +10556,34 @@ bool player::is_wearing_shoes(std::string side) const
...
+        if( i.covers( bp_head ) &&
+            !worn_item->has_flag( "HELMET_COMPAT" ) &&
+            !worn_item->has_flag( "SKINTIGHT" ) &&
+            !worn_item->has_flag( "OVERSIZE" ) ) {
+            ret = true;
```
Comment (Coolthulhu) [inline src/player.cpp]: `return true`, then change the last line in the function to `return false`.

**Pull request merged**